sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -127,6 +194,66 @@ private static PipelineResult run(
     return Linq4j.asEnumerable(values);
   }
 
+  private static Enumerable<Object> limitCollect(PipelineOptions options, BeamRelNode node) {
+    long id = options.getOptionsId();
+    Queue<Object> values = new ConcurrentLinkedQueue<Object>();
+
+    checkArgument(
+        options
+            .getRunner()
+            .getCanonicalName()
+            .equals("org.apache.beam.runners.direct.DirectRunner"));
+
+    LimitStateWrapper stateWrapper = new LimitStateWrapper();
+    LimitCounter.globalValues.put(id, values);
+    LimitCounter.globalLimitArguments.put(id, getLimitCount(node));
+    LimitCounter.globalStates.put(id, stateWrapper);
+    limitRun(options, node, new LimitCounter(), stateWrapper);
+    LimitCounter.globalValues.remove(id);
+    LimitCounter.globalLimitArguments.remove(id);
+    LimitCounter.globalStates.remove(id);
+
+    return Linq4j.asEnumerable(values);
+  }
+
+  private static class LimitCounter extends DoFn<Row, Void> {
+    private static final Map<Long, Integer> globalLimitArguments =
```

Comment:
In the `DirectRunner` this is OK. I think you could pretty easily make it a bit more general by using stateful ParDo(DoFn) for counting. Then the only thing left to make it work on a distributed runner is a communication path back to the shell. If you leave it like this, I think the `DoFn` also is not ready, because in a distributed run there will be lots of copies and separate JVMs on different workers, not gathered to one key, etc.

Reply:
Use stateful ParDo to implement counter/cancelling now.